crates/datasources/src/sqlite/mod.rs

```diff
@@ -0,0 +1,426 @@
+        };
+
+        // Build WHERE clause if predicate pushdown enabled.
+        //
+        // TODO: This may produce an invalid clause. We'll likely only want to
+        // convert some predicates.
+        let predicate_string = {
+            exprs_to_predicate_string(filters)
+                .map_err(|e| DataFusionError::External(Box::new(e)))?
+        };
+
+        // Build copy query.
+        let query = format!(
+            "SELECT {} FROM {} {} {} {}",
+            projection_string, // SELECT <str>
+            self.table,        // <table>
+            // [WHERE]
+            if predicate_string.is_empty() {
+                ""
+            } else {
+                "WHERE "
+            },
+            predicate_string.as_str(), // <where-predicate>
+            limit_string,              // [LIMIT ..]
+        );
+
+        Ok(Arc::new(SqliteQueryExec {
+            query,
+            state: self.state.clone(),
+            schema: projected_schema,
+            metrics: ExecutionPlanMetricsSet::new(),
+        }))
+    }
+}
+
+#[derive(Debug)]
+pub struct SqliteQueryExec {
+    query: String,
+    state: SqliteAccessState,
+    schema: SchemaRef,
+    metrics: ExecutionPlanMetricsSet,
+}
+
+impl ExecutionPlan for SqliteQueryExec {
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn schema(&self) -> SchemaRef {
+        self.schema.clone()
+    }
+
+    fn output_partitioning(&self) -> Partitioning {
+        Partitioning::UnknownPartitioning(1)
+    }
+
+    fn output_ordering(&self) -> Option<&[PhysicalSortExpr]> {
+        None
+    }
+
+    fn children(&self) -> Vec<Arc<dyn ExecutionPlan>> {
+        Vec::new()
+    }
+
+    fn with_new_children(
+        self: Arc<Self>,
+        _children: Vec<Arc<dyn ExecutionPlan>>,
+    ) -> DataFusionResult<Arc<dyn ExecutionPlan>> {
+        Err(DataFusionError::Execution(
+            "cannot replace children for SqliteQueryExec".to_string(),
+        ))
+    }
+
+    fn execute(
+        &self,
+        partition: usize,
+        _context: Arc<TaskContext>,
+    ) -> DataFusionResult<SendableRecordBatchStream> {
+        if partition != 0 {
+            return Err(DataFusionError::Execution(format!(
+                "invalid partition: {partition}"
+            )));
+        }
+
+        let state = self.state.clone();
+        let query = self.query.clone();
+        let schema = self.schema.clone();
+
+        let stream = async_stream::stream! {
+            let mut stream = state.client.query(query, []);
+            let conv = Converter::new(schema);
+
+            while let Some(batch) = stream.next().await {
+                let data = match batch {
+                    Ok(b) => b.data,
+                    Err(e) => {
+                        yield Err(DataFusionError::Execution(e.to_string()));
+                        return;
+                    }
+                };
+                match conv.create_record_batch(data) {
```

Comment:
how big are the batches we're getting? does it matter?

Reply:
each batch is at-most 1000 rows

Reply:
I'd add a comment just to make that clear to would-be-users.

Reply:
We have:

```
const BATCH_SIZE: usize = 1000;
```

in `async_sqlite.rs`